casanovo/denovo/model.py

```diff
@@ -460,6 +928,52 @@ def on_predict_epoch_end(
                         ),
                     )
 
+    def _get_output_peptide_and_scores(
+        self, aa_tokens: List[str], aa_scores: torch.Tensor
+    ) -> Tuple[str, List[str], float, str]:
+        """
+        Get peptide to output, amino acid and peptide-level confidence scores.
+
+        Parameters
+        ----------
+        aa_tokens : List[str]
+            Amino acid tokens of the peptide sequence.
+        aa_scores : torch.Tensor
+            Amino acid-level confidence scores for the predicted sequence.
+
+        Returns
+        -------
+        peptide : str
+            Peptide sequence.
+        aa_tokens : List[str]
+            Amino acid tokens of the peptide sequence.
+        peptide_score : str
+            Peptide-level confidence score.
+        aa_scores : str
+            Amino acid-level confidence scores for the predicted sequence.
+        """
+        # Omit stop token.
+        aa_tokens = aa_tokens[1:] if self.decoder.reverse else aa_tokens[:-1]
+        peptide = "".join(aa_tokens)
```

Comment:
**question:** One final confusion about the order of the tokens and AAs. Here we create `peptide` from `aa_tokens` without reversing them, whereas in line 969 the tokens are reversed prior to getting the AA scores. Doesn't that contradict itself?
Although I guess the scores are reversed twice in lines 969-970, but why do `aa_scores` only need the be reversed once?

Reply:
`aa_tokens` arrives in the correct order in any case (except for the position of stop token) but aa scores arrive reversed `if self.decoder.reverse = True` so that's why we a) reverse `aa_tokens` to correspond to the order of scores for retrieval and then b) retrieved aa scores are reversed to correspond to the N- to C-terminus order. Let me know if I'm missing anything.

Reply:
Ok, rather than reversing `aa_tokens` twice, wouldn't it be easier to reverse `aa_scores` only once? E.g.:
```
top_aa_scores = [
    aa_score[self.decoder._aa2idx[aa_token]].item()
    for aa_score, aa_token in zip(aa_scores[::step], aa_tokens)
]
```

Slightly less code, but also less confusing to understand what's happening I think.

Reply:
`aa_scores` is a tensor with size `(max_length, vocab)` so both the snippet you shared doesn't work as is and also reversal of the non-empty positions in the tensor wasn't as straightforward for me.

Reply:
That makes total sense! Thanks.